### DISCUSSION TITLE
Automatic Updates

### DISCUSSION THREAD

**Discussion by bencroker** [craftcms/cms#12868] [Ideas] [answered]
This may already be on your radar, but Drupal is working towards the ability to enable [automatic updates](https://www.drupal.org/project/ideas/issues/2940731) (currently via [this module](https://www.drupal.org/project/automatic_updates). A few notable points about the approach:

- Applies patch-level updates only (`Z` in `X.Y.Z`).
- Allows updates to be run on a copy of the codebase using [Composer Stager](https://github.com/php-tuf/composer-stager), meaning less time in maintenance mode.
- Currently applies to core only, not to modules.
- Requires a scheduled cron job.

Assuming that patch updates in Craft never contain breaking changes (which is already the case) or database migrations then this would be a safe way to ensure that security and bug fixes are applied as soon as possible after being released and reduce the maintenance burden on developers/agencies. The danger, of course, is that something goes wrong, for whatever reason, and the site goes down. To help avoid this, Drupal is adding “readiness checks” as well as the signing of releases according to [The Update Framework](https://theupdateframework.io/) (which may be or may not be related).

Wordpress, as most people know, has supported [automatic updates](https://wordpress.org/documentation/article/configuring-automatic-background-updates/) for a long time but they are run via [WP-Cron](https://developer.wordpress.org/plugins/cron/) which just sounds like the worst idea ever.

**Comment by brandonkelly** [comment, marked as answer]
I don’t think your production-installed Composer dependencies should ever be able to get a life of their own and diverge from what’s in your repo.

That said, I could see us doing something with automatic update PRs (and providing a GitHub action that could be used to automatically merge those PRs if you’re comfortable with it).

**Reply by brandonkelly** [reply]
We can’t be certain that a patch update isn’t going to contain new migrations though, and migrations that modify the project config _must_ be run on an environment that allows project config changes. And any project config changes must be committed to the repo along with the updated `composer.lock` file.

So we would have two options:

- Only PR updates that don’t make any changes to the `src/migrations/` folder, and/or where the schema version doesn’t change
- Make this a Craft Cloud-only feature, where we have the ability to spin up a temp environment with the production DB, run new migrations, and then PR the updated project config changes along with the updated `composer.lock` file.

Considering it would be a little costly to run this service either way, it probably makes the most sense to be thinking of this as a Cloud feature anyway.

**Reply by bencroker** [reply]
Agreed, the production composer file getting out of sync with the repo is kind of a deal-breaker here. The automatic update PRs idea sounds very interesting though. I’m imagining something like [Automating Dependabot with GitHub Actions](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/automating-dependabot-with-github-actions) that only runs (or deploys) on patch updates that don’t contain migrations. This may even already be possible with the existing [configuration options for `dependabot.yml`](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file) combined with a custom GitHub action. I’ll make a note to look into it, thanks for the response!

**Reply by brandonkelly** [reply]
Ah yeah, didn’t think of Dependabot! I’d still be a little hesitant there unless there’s some way to check that the update doesn’t touch the `src/migrations/` folder.

**Reply by bencroker** [reply]
Yup, the GitHub action would have to check whether the update contains any migrations and only run if negative.
